pageserver/src/storage_sync.rs

```diff
@@ -129,1743 +150,722 @@
-                            conf,
-                            (storage.as_ref(), &index, sync_queue),
-                            current_remote_timeline.as_ref(),
-                            sync_id,
-                            new_download_data,
-                            sync_start,
-                            "download",
-                        )
-                        .await;
-                    }
-                    ControlFlow::Break(_) => {
-                        index
-                            .write()
-                            .await
-                            .set_awaits_download(&sync_id, false)
-                            .ok();
-                    }
+                UploadOp::UploadMetadata(_, _) => {
+                    upload_queue.num_inprogress_metadata_uploads += 1;
                 }
-            }
-            DownloadMarker::Nothing
-        }
-        .instrument(info_span!("download_timeline_data")),
-    );
-
-    if let Some(mut delete_data) = batch.delete {
-        if upload_result.is_some() {
-            match validate_task_retries(delete_data, max_sync_errors)
-                .instrument(info_span!("retries_validation"))
-                .await
-            {
-                ControlFlow::Continue(new_delete_data) => {
-                    delete_timeline_data(
-                        conf,
-                        (storage.as_ref(), &index, sync_queue),
-                        sync_id,
-                        new_delete_data,
-                        sync_start,
-                        "delete",
-                    )
-                    .instrument(info_span!("delete_timeline_data"))
-                    .await;
+                UploadOp::Delete(_) => {
+                    upload_queue.num_inprogress_deletions += 1;
+                }
+                UploadOp::Barrier(sender) => {
+                    sender.send_replace(());
+                    continue;
                 }
-                ControlFlow::Break(failed_delete_data) => {
-                    if let Err(e) = update_remote_data(
-                        conf,
-                        storage.as_ref(),
-                        &index,
-                        sync_id,
-                        RemoteDataUpdate::Delete(&failed_delete_data.data.deleted_layers),
+            };
+
+            // Assign unique ID to this task
+            upload_queue.task_counter += 1;
+            let task_id = upload_queue.task_counter;
+
+            // Add it to the in-progress map
+            let task = Arc::new(UploadTask {
+                task_id,
+                op: next_op,
+                retries: AtomicU32::new(0),
+            });
+            upload_queue
+                .inprogress_tasks
+                .insert(task.task_id, Arc::clone(&task));
+
+            // Spawn task to perform the task
+            let self_rc = Arc::clone(self);
+            let _handle = self.runtime.spawn(async move {
```

Comment:
If we loose the handle, we would never know if the task had finished or not, it could panic, for example.
Channels could also help to override this, but IMO we should not leave such things unaccounted for "endless" or potentially endless tasks such as this one.

We also won't be able to cancel such tasks properly without dropping the runtime: `wait_completion` could hang for a lot of time before reaches the barrier it was asked to reach for, not sure it's a good behaviour either: if some of the uploads fails during that waiting, we won't update metadata hence just wait a lot of time for no good increment.